### PR TITLE
Feature/add multiple queues

### PR DESCRIPTION
**Pull request by nettashviro**
**ADDED**

- support in multiple queues and middlewares
- only one connection to collection
- multiple msgs for a queue
-  when updating return the full-document
- added rabbit status function
-  add an option to increase the number of retries reconnect to rabbit
- add mongo reconnection and mongo status

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,18 +38,27 @@ contains 3 fields:
...
 | 1 | `silent`  | `boolean` | if false, logs connection and changes to the console | `true` |
 | 2 | `prettify`  | `boolean` | if true, will filter the result and send it in a specific format | `true`|
-| 3 | `middleware`  | `(DataObjectType) => (string \| Object \| Buffer)` | will only work with `prettify:true`. A function for manipulating the prettified data received from the listener before sending it to the queue | [identity function](https://en.wikipedia.org/wiki/Identity_function) |
+| 3 | `retries`  | `number` | amount of retries to connect to rabbit | `5`| 
+
+
+
 
+### Types:
+| #  | field | type | info | default |
```
Comment (Shahar-Y) [inline README.md:47]: What is the default?

File: README.md [hunk 2]
```diff
@@ -38,18 +38,27 @@ contains 3 fields:
...
+|---|---|---|---|---|
+|1| `queueObjectType`| `name: string, middleware?: middlewareFunc`| queue name and the middleware parser
+|2|`middlewareFunc` | `(DataObjectType, collectionName) => (null | string | Object | Buffer | string[] | Object[] | Buffer[] | undefined)` | A function for manipulating the prettified data received from the listener before sending it to the queue. will only work with  a `prettify:true`. | [identity function](https://en.wikipedia.org/wiki/Identity_function)
+
+___
 #### Example:
```
Comment (Shahar-Y) [inline README.md:54]: Is the example still relevant?

File: src/paramTypes.ts [hunk 3]
```diff
@@ -18,8 +18,18 @@ export type DataObjectType = {
 	};
 }
 
+export type queueObjectType = {
+	name: string,
+	middleware?: middlewareFunc, 
+};
+
+export type middlewareFunc =
```
Comment (Shahar-Y) [inline src/paramTypes.ts]:
```suggestion
export type MiddlewareFuncType = 
```
types should begin with upper-case and end with 'Type'

File: src/paramTypes.ts [hunk 4]
```diff
@@ -18,8 +18,18 @@ export type DataObjectType = {
 	};
 }
 
+export type queueObjectType = {
```
Comment (Shahar-Y) [inline src/paramTypes.ts]:
```suggestion
export type QueueObjectType = {
```

File: src/example/connect.ts [hunk 5]
```diff
@@ -1,37 +1,38 @@
...
 // Second colQCouple
 const mongoData2: MongoDataType = {
-	collectionName: 'files',
+	collectionName: 'permissions',
 	connectionString: 'mongodb://localhost:27017/devDB?replicaSet=rs0',
 };
 
 const rabbitData2: RabbitDataType = {
-	queueName: 'MyQueueName2',
+	queues: [{ name:'Queue3'}],
```
Comment (Shahar-Y) [inline src/example/connect.ts]: 3?

File: src/index.ts [hunk 6]
```diff
@@ -26,25 +40,88 @@ const defaultOptions: MTROptions = {
...
+    while(true) {
+        try {
+            log('try connect to mongo', options);
+            await mongoConn.connect();
+            log(`successful connection to mongo on connectionString: ${mongoData.connectionString}`, options);
+            initWatch(mongoData,rabbitData, options);
+            break;
+        } catch(error) {
+            log('cant connect to mongo', options);
+            await sleep(10000);
```
Comment (Shahar-Y) [inline src/index.ts]: Make it const

File: src/index.ts [hunk 7]
```diff
@@ -26,25 +40,88 @@ const defaultOptions: MTROptions = {
...
+async function mongoConnection(mongoData: MongoDataType, rabbitData: RabbitDataType, options: MTROptions) {
+    while(true) {
+        try {
+            log('try connect to mongo', options);
+            await mongoConn.connect();
+            log(`successful connection to mongo on connectionString: ${mongoData.connectionString}`, options);
+            initWatch(mongoData,rabbitData, options);
+            break;
+        } catch(error) {
+            log('cant connect to mongo', options);
```
Comment (Shahar-Y) [inline src/index.ts]:
```suggestion
            log('cant connect to mongo. Retrying in ${reconnectSleepTime}', options);
```

File: src/index.ts [hunk 8]
```diff
@@ -26,25 +40,88 @@ const defaultOptions: MTROptions = {
...
+ * @param mongoData  - mongo data - collections and connection uri (MongoDataType)
+ * @param rabbitData - rabbit data (RabbitDataType)
+ * @param options    - contains the MTROptions.
+ */
+async function mongoConnection(mongoData: MongoDataType, rabbitData: RabbitDataType, options: MTROptions) {
+    while(true) {
+        try {
+            log('try connect to mongo', options);
+            await mongoConn.connect();
+            log(`successful connection to mongo on connectionString: ${mongoData.connectionString}`, options);
```
Comment (Shahar-Y) [inline src/index.ts:81]: maybe make these console.log, so it will always print upon connection / error?

File: src/index.ts [hunk 9]
```diff
@@ -53,25 +130,43 @@ export default async function watchAndNotify(mongoData: MongoDataType, rabbitDat
...
+
+    // Define change stream
+    const pipeline = [{ $match: { 'ns.db': connectionObject.endpoint, 'ns.coll': mongoData.collectionName } }];
+    const optionsStream : ChangeStreamOptions = {fullDocument: 'updateLookup'}
+    const changeStream = collection.watch(pipeline, optionsStream);
+
+    // start listen to changes
+    changeStream.on('change', (event: mongodb.ChangeEvent<Object>) => {
+        rabbitData.queues.forEach((queue) => {
+            const formattedData = options.prettify ?
```
Comment (Shahar-Y) [inline src/index.ts]: :+1:

File: src/paramTypes.ts [hunk 10]
```diff
@@ -18,8 +18,18 @@ export type DataObjectType = {
...
+	(dataObject: DataObjectType, collection: string) => 
+	(null| string | Object | Buffer |string[] | Object[] 
+		| Buffer[] | undefined);
+
 export type MTROptions = {
 	silent: boolean;
 	prettify: boolean;
-	middleware: (datqaObject: DataObjectType) => (string | Object | Buffer);
-}
+	retries?: number;
```
Comment (Shahar-Y) [inline src/paramTypes.ts]: if the retries are about the connection to rabbit, perhaps call them rabbitRetries?

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Great work. Have you tried testing it with disconnections and other types of plausible errors?

**Comment by nettashviro** [comment]
> Great work. Have you tried testing it with disconnections and other types of plausible errors?

of course! I checked all sorts of disconnection issues - both in Mongo and rabbitmq, and various edge cases.

File: src/paramTypes.ts (hunk 10 above)
Comment (nettashviro) [reply src/paramTypes.ts]: done

File: src/index.ts (hunk 7 above)
Comment (nettashviro) [reply src/index.ts]: done

File: src/index.ts (hunk 6 above)
Comment (nettashviro) [reply src/index.ts]: done

File: src/paramTypes.ts (hunk 4 above)
Comment (nettashviro) [reply src/paramTypes.ts]: done

File: src/paramTypes.ts (hunk 3 above)
Comment (nettashviro) [reply src/paramTypes.ts]: done

File: src/example/connect.ts (hunk 5 above)
Comment (nettashviro) [reply src/example/connect.ts]: In the previous example I set up two queues, so this is queue 3. Anyway, I changed this back to queue2 in any case this is confusing.

File: README.md (hunk 2 above)
Comment (nettashviro) [reply README.md:54]: yep

File: README.md (hunk 1 above)
Comment (nettashviro) [reply README.md:47]: there is not default for Queue-Type, the user **must** define a name for the queue

File: src/index.ts (hunk 8 above)
Comment (nettashviro) [reply src/index.ts:81]: done

File: src/paramTypes.ts [hunk 11]
```diff
@@ -18,8 +20,30 @@ export type DataObjectType = {
...
+	routingKey?: string,
+}
+
+export type QueueObjectType = {
+	name: string,
+	exchange?: ExchangeObjectType,
+	middleware?: MiddlewareFuncType, 
+};
+
+export type sendMsgReq = {
```
Comment (Shahar-Y) [inline src/paramTypes.ts]:
```suggestion
export type SendMsgReqType = {
```

File: src/index.ts [hunk 12]
```diff
@@ -53,25 +137,45 @@ export default async function watchAndNotify(mongoData: MongoDataType, rabbitDat
...
+    }).on('error', async(err) => {
+        console.log(`error in mongo`);
+        console.log(err);
+        mongoConnection(mongoData, rabbitData, options);
     });
+
+    console.log(`MTR: ===> successful connection to collection: ${mongoData.collectionName}`);
+}
+
+export function sendMsg(queue: QueueObjectType, msg: any) {
```
Comment (Shahar-Y) [inline src/index.ts]: Add comments

File: src/index.ts [hunk 13]
```diff
@@ -26,25 +41,94 @@ const defaultOptions: MTROptions = {
...
+    
+    mongoConn = new MongoClient(mongoData.connectionString, {useUnifiedTopology: true});
+    mongoConnection(mongoData, rabbitData, options);
+}
+
+function sleep(ms: number) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function mongoConnection(mongoData: MongoDataType, rabbitData: RabbitDataType, options: MTROptions) {
```
Comment (Shahar-Y) [inline src/index.ts:74]: Add comments

File: src/index.ts [hunk 14]
```diff
@@ -1,20 +1,35 @@
...
+import { MongoClient } from 'mongodb';
 
+let mongoConn: MongoClient;
 const csParser = new CSParser({
     scheme: 'mongodb',
     hosts: []
 });
 
+const defaultMiddleware: MiddlewareFuncType = (data: DataObjectType) => {
+    return {data};
```
Comment (Shahar-Y) [inline src/index.ts]: Why does it encapsulate the data inside of another object?

File: src/paramTypes.ts [hunk 15]
```diff
@@ -1,11 +1,13 @@
+import { ExchangeType } from 'menashmq';
```
Comment (Shahar-Y) [inline src/paramTypes.ts:1]: I'd add comments about the types explaining their meaning. Maybe even from the readme

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Good job. Some small fixes required

File: src/index.ts (hunk 12 above)
Comment (nettashviro) [reply src/index.ts]: done

File: src/index.ts (hunk 13 above)
Comment (nettashviro) [reply src/index.ts:74]: done

File: src/index.ts (hunk 14 above)
Comment (nettashviro) [reply src/index.ts]: My bad, it was an object at first and I forgot to delete the curly brackets, nice catch!

File: src/paramTypes.ts (hunk 15 above)
Comment (nettashviro) [reply src/paramTypes.ts:1]: done, added a link to exchange types information in the readme file

File: src/index.ts [hunk 16]
```diff
@@ -58,10 +44,33 @@ export default async function watchAndNotify(mongoData: MongoDataType, rabbitDat
...
 }
 
+
 function sleep(ms: number) {
     return new Promise(resolve => setTimeout(resolve, ms));
 }
 
+/**
+ * Get rabbit health status
+ * @returns {boolean} - true if healthy
```
Comment (Shahar-Y) [inline src/index.ts]: This seems like it returns a boolean inside of an object :confused:

File: src/index.ts [hunk 17]
```diff
@@ -58,10 +44,33 @@ export default async function watchAndNotify(mongoData: MongoDataType, rabbitDat
...
+ */
+export function getMongoHealthStatus(): boolean {
+    return mongoConn.isConnected();
+}
+
+/**
+ * Start mongo connection
+ * @param {MongoDataType}   mongoData   - mongo uri and collection name
+ * @param {RabbitDataType}  rabbitData  - rabbit data
+ * @param {MTROptions}      options     - mongo to rabbit options
```
Comment (Shahar-Y) [inline src/index.ts:72]: Why all the curly-braces? It it a known convention?

File: src/index.ts [hunk 18]
```diff
@@ -172,15 +154,47 @@ async function initWatch(mongoData: MongoDataType, rabbitData: RabbitDataType, o
...
+            sendMsg(queue, formattedData);
+    }
+}
+/**
+ * sendMsg function to queue - by exchange or direct queue
+ * @param {QueueObjectType} queue   - queue object 
+ * @param {any}             msg     - formatted msg
+ */
+export function sendMsg(queue: QueueObjectType, msg: any) {   
+    console.log('sending msg');
```
Comment (Shahar-Y) [inline src/index.ts]: not informative. remove it or use the log and print a more informative data in the format

**Review by Shahar-Y — CHANGES_REQUESTED** [review]
Looks great

File: src/index.ts (hunk 17 above)
Comment (nettashviro) [reply src/index.ts:72]: yes

File: src/index.ts (hunk 16 above)
Comment (nettashviro) [reply src/index.ts]: it's a known convention

File: src/index.ts (hunk 18 above)
Comment (nettashviro) [reply src/index.ts]: done

**Pull request merged**